// openbim-components/src/annotation/CircleAnnotation/index.js
import { Vector2 } from "three";
import { BaseSVGAnnotation, UIElement } from "../../base-types";
import { Button } from "../../ui";
import { SVGCircle } from "../SVGCircle";
export class CircleAnnotation extends BaseSVGAnnotation {
    constructor(components, drawManager) {
        super(components);
        this.name = "CircleAnnotation";
        this.canvas = null;
        this.uiElement = new UIElement();
        this._cursorPosition = new Vector2();
        this.start = (e) => {
            var _a, _b, _c, _d;
            if (!this.canDraw) {
                return undefined;
            }
            if (!this._isDrawing) {
                this._isDrawing = true;
                this._previewElement.setStyle((_a = this.drawManager) === null || _a === void 0 ? void 0 : _a.viewport.config);
                this._previewElement.cx = e.clientX;
                this._previewElement.cy = e.clientY;
                (_b = this.svgViewport) === null || _b === void 0 ? void 0 : _b.append(this._previewElement.get());
            }
            else {
                const circle = this._previewElement.clone();
                circle.setStyle((_c = this.drawManager) === null || _c === void 0 ? void 0 : _c.viewport.config);
                (_d = this.svgViewport) === null || _d === void 0 ? void 0 : _d.append(circle.get());
                this.cancel();
                return circle;
            }
            return undefined;
        };
        this.cancel = () => {
            if (!this._isDrawing) {
                return;
            }
            this._isDrawing = false;
            this._previewElement.reset();
            this._previewElement.get().remove();
        };
        this.draw = (e) => {
            if (!this.canDraw || !this._isDrawing) {
                return;
            }
            this._cursorPosition.x = e.clientX;
            this._cursorPosition.y = e.clientY;
            this._previewElement.radius = this._cursorPosition.distanceTo(this._previewElement.centerPoint);
        };
        this._previewElement = new SVGCircle(components);
        this.drawManager = drawManager;
        const main = new Button(components);
        this.uiElement.set({ main });
        main.label = "Circle";
        main.materialIcon = "radio_button_unchecked";
        main.onClick.add(() => {
            if (this.drawManager) {
                this.drawManager.activateTool(this);
            }
            else {
                this.enabled = !this.enabled;
            }
        });
    }
    async dispose() {
        await super.dispose();
        this._previewElement.dispose();
    }
}
