// openbim-clay/primitives/Primitive/index.js
import * as THREE from "three";
import { Selector } from "../../utils";
export class Primitive {
    constructor() {
        /**
         * All the selected items within this primitive.
         */
        this.selected = new Selector();
        this._baseColor = new THREE.Color(0.5, 0.5, 0.5);
        this._selectColor = new THREE.Color(1, 0, 0);
        this.list = {};
    }
    /**
     * The list of ids of the {@link list} of items.
     */
    get ids() {
        const ids = [];
        for (const id in this.list) {
            ids.push(this.list[id].id);
        }
        return ids;
    }
    /**
     * The color of all the points.
     */
    get baseColor() {
        return this._baseColor;
    }
    /**
     * The color of all the points.
     */
    set baseColor(color) {
        this._baseColor.copy(color);
    }
    /**
     * The color of all the selected points.
     */
    get selectColor() {
        return this._selectColor;
    }
    /**
     * The color of all the selected points.
     */
    set selectColor(color) {
        this._selectColor.copy(color);
    }
    get _positionBuffer() {
        return this.mesh.geometry.attributes.position;
    }
    get _colorBuffer() {
        return this.mesh.geometry.attributes.color;
    }
    get _normalBuffer() {
        return this.mesh.geometry.attributes.normal;
    }
    get _attributes() {
        return Object.values(this.mesh.geometry.attributes);
    }
}
