// openbim-components/src/ifc/IfcPropertiesManager/src/prop-actions.js
import { SimpleUIComponent, FloatingWindow, TextInput, Button, } from "../../../ui";
import { Event } from "../../../base-types";
import { IfcPropertiesUtils } from "../../IfcPropertiesUtils";
export class PropActionsUI extends SimpleUIComponent {
    constructor(components) {
        const div = document.createElement("div");
        div.className = "flex";
        super(components, `<div class="flex"></div>`);
        this.modalVisible = false;
        this.onEditProp = new Event();
        this.onRemoveProp = new Event();
        this.data = {};
        this.editPropBtn = new Button(this._components);
        this.editPropBtn.materialIcon = "edit";
        this.setEditUI();
        this.removePropBtn = new Button(this._components);
        this.removePropBtn.materialIcon = "delete";
        this.setRemoveUI();
        this.addChild(this.editPropBtn, this.removePropBtn);
        this._modal = new SimpleUIComponent(components, `<dialog></dialog>`);
        this._components.ui.add(this._modal);
        this._modal.get().addEventListener("close", () => {
            this.removeFromParent();
            this.modalVisible = false;
            this._modalWindow.visible = false;
        });
        this._modalWindow = new FloatingWindow(this._components);
        this._modalWindow.get().className =
            "overflow-auto text-white bg-ifcjs-100 rounded-md w-[350px]";
        this._modalWindow.onHidden.add(() => this._modal.get().close());
        this._modal.addChild(this._modalWindow);
    }
    async dispose(onlyChildren = false) {
        await super.dispose(onlyChildren);
        this.onRemoveProp.reset();
        await this.editPropBtn.dispose();
        await this.removePropBtn.dispose();
        await this._modal.dispose();
        await this._modalWindow.dispose();
        this.data = {};
    }
    setEditUI() {
        const editUI = new SimpleUIComponent(this._components, `<div class="flex flex-col gap-y-4 p-4 overflow-auto"></div>`);
        const nameInput = new TextInput(this._components);
        nameInput.label = "Name";
        const valueInput = new TextInput(this._components);
        valueInput.label = "Value";
        const acceptBtn = new Button(this._components);
        acceptBtn.materialIcon = "check";
        acceptBtn.label = "Accept";
        acceptBtn.get().classList.remove("hover:bg-ifcjs-200");
        acceptBtn.get().classList.add("hover:bg-success");
        acceptBtn.onClick.add(async () => {
            this._modal.get().close();
            const { model, expressID } = this.data;
            if (!model || !expressID)
                return;
            await this.onEditProp.trigger({
                model,
                expressID,
                name: nameInput.value,
                value: valueInput.value,
            });
        });
        const cancelBtn = new Button(this._components);
        cancelBtn.materialIcon = "close";
        cancelBtn.label = "Cancel";
        cancelBtn.get().classList.remove("hover:bg-ifcjs-200");
        cancelBtn.get().classList.add("hover:bg-error");
        cancelBtn.onClick.add(() => this._modal.get().close());
        const actionBtns = new SimpleUIComponent(this._components, `<div class="flex gap-x-2 justify-end"></div>`);
        actionBtns.addChild(cancelBtn, acceptBtn);
        editUI.addChild(nameInput, valueInput, actionBtns);
        this.editPropBtn.onClick.add(async () => {
            var _a, _b, _c, _d, _e, _f, _g, _h;
            const { model, expressID } = this.data;
            const properties = model === null || model === void 0 ? void 0 : model.properties;
            if (!model || !expressID || !properties)
                return;
            const prop = properties[expressID];
            const { key: nameKey } = IfcPropertiesUtils.getEntityName(properties, expressID);
            if (nameKey) {
                nameInput.value = (_b = (_a = prop[nameKey]) === null || _a === void 0 ? void 0 : _a.value) !== null && _b !== void 0 ? _b : "";
            }
            else {
                nameInput.value = (_d = (_c = prop.Name) === null || _c === void 0 ? void 0 : _c.value) !== null && _d !== void 0 ? _d : "";
            }
            const { key: valueKey } = IfcPropertiesUtils.getQuantityValue(properties, expressID);
            if (valueKey) {
                valueInput.value = (_f = (_e = prop[valueKey]) === null || _e === void 0 ? void 0 : _e.value) !== null && _f !== void 0 ? _f : "";
            }
            else {
                valueInput.value = (_h = (_g = prop.NominalValue) === null || _g === void 0 ? void 0 : _g.value) !== null && _h !== void 0 ? _h : "";
            }
            this._modalWindow.title = "Edit Property";
            this._modalWindow.setSlot("content", editUI);
            this.showModal();
        });
    }
    setRemoveUI() {
        const removeUI = new SimpleUIComponent(this._components, `<div class="flex flex-col gap-y-4 p-4 overflow-auto"></div>`);
        const warningText = document.createElement("div");
        warningText.className = "text-base text-center";
        warningText.textContent =
            "Are you sure to delete this property? This action can't be undone.";
        removeUI.get().append(warningText);
        const acceptBtn = new Button(this._components);
        acceptBtn.materialIcon = "check";
        acceptBtn.label = "Accept";
        acceptBtn.get().classList.remove("hover:bg-ifcjs-200");
        acceptBtn.get().classList.add("hover:bg-success");
        acceptBtn.onClick.add(async () => {
            this._modal.get().close();
            const { model, expressID, setID } = this.data;
            if (!model || !expressID || !setID)
                return;
            this.removeFromParent(); // As the psetUI is going to be disposed, then we need to first remove the action buttons so they do not become disposed as well.
            await this.onRemoveProp.trigger({ model, expressID, setID });
        });
        const cancelBtn = new Button(this._components);
        cancelBtn.materialIcon = "close";
        cancelBtn.label = "Cancel";
        cancelBtn.get().classList.remove("hover:bg-ifcjs-200");
        cancelBtn.get().classList.add("hover:bg-error");
        cancelBtn.onClick.add(() => this._modal.get().close());
        const actionBtns = new SimpleUIComponent(this._components, `<div class="flex gap-x-2 justify-end"></div>`);
        actionBtns.addChild(cancelBtn, acceptBtn);
        removeUI.addChild(actionBtns);
        this.removePropBtn.onClick.add(async () => {
            const { model, expressID, setID } = this.data;
            if (!model || !expressID || !setID)
                return;
            this._modalWindow.title = "Remove Property";
            this._modalWindow.setSlot("content", removeUI);
            this.showModal();
        });
    }
    showModal() {
        this.modalVisible = true;
        this._modalWindow.visible = true;
        this._modal.get().showModal();
    }
}
