// bim-fragment/src/flatbuffers/fragments/index/fragments.js
// automatically generated by the FlatBuffers compiler, do not modify
import * as flatbuffers from 'flatbuffers';
import { Fragment } from '../../fragments/index/fragment.js';
export class Fragments {
    constructor() {
        this.bb = null;
        this.bb_pos = 0;
    }
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsFragments(bb, obj) {
        return (obj || new Fragments()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsFragments(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new Fragments()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    items(index, obj) {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? (obj || new Fragment()).__init(this.bb.__indirect(this.bb.__vector(this.bb_pos + offset) + index * 4), this.bb) : null;
    }
    itemsLength() {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    static startFragments(builder) {
        builder.startObject(1);
    }
    static addItems(builder, itemsOffset) {
        builder.addFieldOffset(0, itemsOffset, 0);
    }
    static createItemsVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addOffset(data[i]);
        }
        return builder.endVector();
    }
    static startItemsVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static endFragments(builder) {
        const offset = builder.endObject();
        return offset;
    }
    static finishFragmentsBuffer(builder, offset) {
        builder.finish(offset);
    }
    static finishSizePrefixedFragmentsBuffer(builder, offset) {
        builder.finish(offset, undefined, true);
    }
    static createFragments(builder, itemsOffset) {
        Fragments.startFragments(builder);
        Fragments.addItems(builder, itemsOffset);
        return Fragments.endFragments(builder);
    }
}
