// openbim-components/src/annotation/RectangleAnnotation/index.js
import { Vector2 } from "three";
import { BaseSVGAnnotation, UIElement } from "../../base-types";
import { Button } from "../../ui";
import { SVGRectangle } from "../SVGRectangle";
export class RectangleAnnotation extends BaseSVGAnnotation {
    constructor(components, drawManager) {
        super(components);
        this.name = "RectangleAnnotation";
        this.canvas = null;
        this.uiElement = new UIElement();
        this._startPoint = new Vector2();
        this.start = (e) => {
            var _a, _b, _c, _d;
            if (!this.canDraw) {
                return undefined;
            }
            if (!this._isDrawing) {
                this._isDrawing = true;
                this._previewElement.setStyle((_a = this.drawManager) === null || _a === void 0 ? void 0 : _a.viewport.config);
                this._startPoint.set(e.clientX, e.clientY);
                (_b = this.svgViewport) === null || _b === void 0 ? void 0 : _b.append(this._previewElement.get());
            }
            else {
                const rectangle = this._previewElement.clone();
                rectangle.setStyle((_c = this.drawManager) === null || _c === void 0 ? void 0 : _c.viewport.config);
                (_d = this.svgViewport) === null || _d === void 0 ? void 0 : _d.append(rectangle.get());
                this.cancel();
                return rectangle;
            }
            return undefined;
        };
        this.cancel = () => {
            if (!this._isDrawing) {
                return;
            }
            this._isDrawing = false;
            this._startPoint.x = 0;
            this._startPoint.y = 0;
            this._previewElement.reset();
            this._previewElement.get().remove();
        };
        this.draw = (e) => {
            if (!this.canDraw || !this._isDrawing) {
                return;
            }
            this._previewElement.x1 = this._startPoint.x;
            this._previewElement.y1 = this._startPoint.y;
            this._previewElement.x2 = e.clientX;
            this._previewElement.y2 = e.clientY;
        };
        this._previewElement = new SVGRectangle(components);
        this.drawManager = drawManager;
        const main = new Button(components);
        this.uiElement.set({ main });
        main.label = "Rectangle";
        main.materialIcon = "crop_square";
        main.onClick.add(() => {
            if (this.drawManager) {
                this.drawManager.activateTool(this);
            }
            else {
                this.enabled = !this.enabled;
            }
        });
    }
    async dispose() {
        await super.dispose();
        this._previewElement.dispose();
    }
}
