// openbim-components/src/navigation/ViewpointsManager/index.js
import { Vector3 } from "three";
import { generateUUID } from "three/src/math/MathUtils";
import { SimpleDimensionLine, LengthMeasurement, } from "../../measurement/LengthMeasurement";
import { Component, Event, UIElement, } from "../../base-types";
import { Button, SimpleUICard, FloatingWindow } from "../../ui";
import { FragmentHighlighter } from "../../fragments";
export class ViewpointsManager extends Component {
    constructor(components) {
        super(components);
        this.name = "ViewpointsManager";
        this.uiElement = new UIElement();
        this.enabled = true;
        this.list = [];
        this.selectionHighlighter = "";
        this.onViewpointViewed = new Event();
        this.onViewpointAdded = new Event();
        this.components = components;
    }
    initialize(config) {
        this.selectionHighlighter = config.selectionHighlighter;
        // this._fragmentGrouper = config.fragmentGrouper;
        // this._fragmentManager = config.fragmentManager;
        this._drawManager = config.drawManager;
        if (this.components.uiEnabled) {
            this.setUI();
        }
    }
    setUI() {
        const viewerContainer = this.components.renderer.get().domElement
            .parentElement;
        const window = new FloatingWindow(this.components);
        window.title = "Viewpoints";
        viewerContainer.append(window.get());
        window.visible = false;
        const main = new Button(this.components, {
            materialIconName: "photo_camera",
        });
        const newButton = new Button(this.components, {
            materialIconName: "add",
            name: "New viewpoint",
        });
        const listButton = new Button(this.components, {
            materialIconName: "format_list_bulleted",
            name: "Viewpoints list",
        });
        listButton.onClick.add(() => {
            window.visible = !window.visible;
        });
        main.addChild(listButton, newButton);
        this.uiElement.set({ main, newButton, window });
    }
    get() {
        throw new Error("Method not implemented.");
    }
    add(data) {
        var _a;
        const { title, description } = data;
        if (!title) {
            return undefined;
        }
        const guid = generateUUID().toLowerCase();
        // #region Store dimensions
        const dimensions = [];
        const dimensionsComp = this.components.tools.get(LengthMeasurement);
        const allDimensions = dimensionsComp.get();
        for (const dimension of allDimensions) {
            dimensions.push({ start: dimension.start, end: dimension.end });
        }
        // #endregion
        // #redgion Store selection
        const highlighter = this.components.tools.get(FragmentHighlighter);
        const selection = highlighter.selection[this.selectionHighlighter];
        // #endregion
        // #region Store filter (WIP)
        // const filter = {entities: "IFCBEAM", storeys: "N07"}
        // #endregion
        // #region Store camera position and target
        const camera = this.components.camera;
        const controls = camera.controls;
        const target = new Vector3();
        const position = new Vector3();
        controls.getTarget(target);
        controls.getPosition(position);
        const projection = camera.getProjection();
        // #endregion
        // #region Store annotations
        const annotations = (_a = this._drawManager) === null || _a === void 0 ? void 0 : _a.saveDrawing(guid);
        // #endregion
        const viewpoint = {
            guid,
            title,
            target,
            position,
            selection,
            // filter,
            description,
            dimensions,
            annotations,
            projection,
        };
        // #region UI representation
        const card = new SimpleUICard(this.components, viewpoint.guid);
        card.title = title;
        card.description = description;
        card.domElement.onclick = () => this.view(viewpoint.guid);
        this.uiElement.get("window").addChild(card);
        // #endregion
        this.list.push(viewpoint);
        this.onViewpointAdded.trigger(guid);
        return viewpoint;
    }
    retrieve(guid) {
        return this.list.find((v) => v.guid === guid);
    }
    async view(guid) {
        const viewpoint = this.retrieve(guid);
        if (!viewpoint) {
            return;
        }
        // #region Recover annotations
        if (this._drawManager && viewpoint.annotations) {
            this._drawManager.viewport.clear();
            this._drawManager.enabled = true;
            this._drawManager.viewport.get().append(viewpoint.annotations);
        }
        // #endregion
        // #region Recover dimensions
        const dimensionsComponent = this.components.tools.get(LengthMeasurement);
        viewpoint.dimensions.forEach((data) => {
            const dimension = new SimpleDimensionLine(this.components, {
                start: data.start,
                end: data.end,
                // @ts-ignore
                lineMaterial: dimensionsComponent._lineMaterial,
                // @ts-ignore
                endpoint: dimensionsComponent._endpointMesh,
            });
            dimension.createBoundingBox();
            // @ts-ignore
            dimensionsComponent._dimensions.push(dimension);
        });
        // #endregion
        // #region Recover filtered elements
        // if (viewpoint.filter) {
        //     const filterData = fragments.groups.get(viewpoint.filter)
        //     for (const fragmentID in fragments.list) {
        //         const fragment = fragments.list[fragmentID]
        //         fragment.setVisibility(fragment.items, false)
        //     }
        //     for (const fragmentID in filterData) {
        //         const ids = filterData[fragmentID]
        //         fragments.list[fragmentID]?.setVisibility(ids, true)
        //     }
        // }
        // #endregion
        // Select elements in the viewpoint
        const selection = {};
        for (const fragmentID in viewpoint.selection) {
            selection[fragmentID] = viewpoint.selection[fragmentID];
        }
        const highlighter = this.components.tools.get(FragmentHighlighter);
        await highlighter.highlightByID(this.selectionHighlighter, selection, true);
        // #region Recover camera position & target
        const camera = this.components.camera;
        const controls = camera.controls;
        controls.setLookAt(viewpoint.position.x, viewpoint.position.y, viewpoint.position.z, viewpoint.target.x, viewpoint.target.y, viewpoint.target.z, true);
        await this.onViewpointViewed.trigger(guid);
        // #endregion
    }
}
